### PR TITLE
feat(IDPay): [IODPAY-110] Initiative card redesign

### PR DESCRIPTION
**Pull request by forrest57**
## Short description
refactored `InitiativeDetailsScreen` and `InitiativeCardComponent` to follow their new design.

Screenshot 2023-03-20 at 17 52 00

https://user-images.githubusercontent.com/60693085/226412500-6b9fab51-2136-42ac-bfbd-eca3305ed9fd.mp4

## List of changes proposed in this pull request
- `InitiativeDetailsScreen`
    - updated styling 
    - added a `lastUpdated` text displaying the timeline's last update
- `InitiativeCardComponent`
    - removed old card background in favor of the new one
    - moved from an absolute positioning approach to a standard Flexbox
    - added required I18n keys
    - refactored styles to match changes
    - added a "percentage of remaining bonus" display slider
    - added conditional labels and text based on bonus status
     - added slot for bonus' image
     - added animated slider under the available amount, displaying the remaining percentage compared to the starting amount

## How to test
navigate to a configured initiative's detail page; the possible statuses are:
- configured but no method/iban:
    - the label will correctly display "no iban" or "no payment method"
    - the numbers and slider will be greyed out
- not configured
    - same as no method/iban, but there will be a different label 
- configured
    - the label will display "active"
    - numbers and slider will be in standard color

### CODE REVIEW

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 1]
```diff
@@ -80,74 +78,90 @@ const InitiativeCardComponent = (props: Props) => {
...
-                !isInitiativeConfigured ? styles.consumedOpacity : {},
-                styles.amount
-              ]}
-            >
-              {formatNumberAmount(amount || 0, true)}
-            </Text>
-            <H5 color="white">
+  const dateString = formatDateAsLocal(endDate, true);
+  const remainingBonusAmountPercentage =
+    (remainingAmount / (amount || remainingAmount)) * 100;
```
Comment (forrest57) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:85]: is this the correct data to be displayed in the "available" field of the card?

**Review by mastro993 — CHANGES_REQUESTED** [review]

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (forrest57) [reply ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:85]: While I do agree this is a cleaner solution, wouldn't amount=0 or amount=undefined be nullish, so default to false?

[the remark this answers is not in the document]
Comment (mastro993) [reply ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx]: Maybe with `amount ? (remainingAmount / amount * 100.0) : 100.0` we can cover all cases

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 2]
```diff
@@ -80,74 +78,92 @@ const InitiativeCardComponent = (props: Props) => {
...
-                !isInitiativeConfigured ? styles.consumedOpacity : {},
-                styles.amount
-              ]}
-            >
-              {formatNumberAmount(amount || 0, true)}
-            </Text>
-            <H5 color="white">
+  const dateString = formatDateAsLocal(endDate, true);
+  const remainingBonusAmountPercentage =
+    amount !== 0 && amount !== undefined
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:85]:
```suggestion
    amount
```

**Comment by mastro993** [comment]
I got some problems:
- Animation is not working
- Amount is going below 0

https://user-images.githubusercontent.com/6160324/224083397-7d007716-5701-4782-b289-483e2008a65b.mp4

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 3]
```diff
@@ -80,74 +83,109 @@ const InitiativeCardComponent = (props: Props) => {
...
+      <VSpacer size={48} />
+
+      {/* bottom part */}
+      <View style={styles.bottomCardSection}>
+        <View style={styles.alignCenter}>
+          <LabelSmall color="bluegreyDark" weight="Regular">
+            {I18n.t("idpay.initiative.details.initiativeCard.availableAmount")}
+          </LabelSmall>
+          <H1 style={!isInitiativeConfigured ? styles.consumedOpacity : {}}>
+            {formatNumberAmount((amount || 0) - (refunded || 0), true)}
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:161]:
```suggestion
            {formatNumberAmount(remainingAmount, true)}
```

**Review by mastro993 — CHANGES_REQUESTED** [review]

File: ts/features/idpay/initiative/details/screens/InitiativeDetailsScreen.tsx [hunk 4]
```diff
@@ -162,6 +166,7 @@ export const InitiativeDetailsScreen = () => {
               )}
             </View>
           </View>
+          <VSpacer size={32} />
         </ScrollView>
         {initiativeNeedsConfiguration && (
           <FooterWithButtons
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/screens/InitiativeDetailsScreen.tsx:172]: There is a small white gap in the lower part of the screen caused by the parent `SafeAreView`.
I'd suggest to remove the parent `SafeAreaView` and instead wrap with it the `FooterWithButtons` component.

**Review by mastro993 — CHANGES_REQUESTED** [review]

File: ts/features/idpay/initiative/details/screens/InitiativeDetailsScreen.tsx [hunk 5]
```diff
@@ -182,18 +187,15 @@ export const InitiativeDetailsScreen = () => {
 
   return (
     <BaseScreenComponent
-      dark={true}
-      titleColor={"white"}
       goBack={true}
-      headerTitle={initiativeData?.initiativeName}
-      headerBackgroundColor={IOColors.bluegrey}
+      headerBackgroundColor={IOColors["blueItalia-50"]}
     >
```
[the remark this answers is not in the document]
Comment (forrest57) [reply ts/features/idpay/initiative/details/screens/InitiativeDetailsScreen.tsx]: adressed in [5073c8c](https://github.com/pagopa/io-app/pull/4436/commits/5073c8cf22c8748069a9828e1b1a1b6bba68857f)

**Comment by thisisjp** [comment]
@forrest57 that’s dope 💦 🚀  just a small chance request: the bar chart animation should be reversed (starting from 100 to the current value), since it represents a value that can only be “emptied” 🙏

**Comment by thisisjp** [comment]
the € currency glyph should be placed after the amount (you might find some inconsistencies in the app, but we’re trying to respect this rule from now on)

File: locales/it/index.yml [hunk 6]
```diff
@@ -3337,8 +3337,15 @@ idpay:
...
-        toRefund: da rimborsare
+        validUntil: Fino al {{expiryDate}}
+        expiresOn: Scade il {{expiryDate}}
+        toRefund: Da rimborsare
+        statusLabels:
+          REFUNDABLE: ATTIVO
+          NOT_REFUNDABLE: DISATTIVATO
+          NOT_REFUNDABLE_ONLY_IBAN: METODO MANCANTE
+          NOT_REFUNDABLE_ONLY_INSTRUMENT: IBAN MANCANTE
+          UNSUBSCRIBED: DISISCRITTO
```
Comment (mastro993) [inline locales/it/index.yml:3348]:
```suggestion
          REFUNDABLE: Attiva
          NOT_REFUNDABLE: Disattivata
          NOT_REFUNDABLE_ONLY_IBAN: Metodo mancante
          NOT_REFUNDABLE_ONLY_INSTRUMENT: IBAN mancante
          UNSUBSCRIBED: Disiscritto
```

File: locales/en/index.yml [hunk 7]
```diff
@@ -3313,8 +3313,15 @@ idpay:
...
-        toRefund: da rimborsare
+        validUntil: Fino al {{expiryDate}}
+        expiresOn: Scade il {{expiryDate}}
+        toRefund: Da rimborsare
+        statusLabels:
+          REFUNDABLE: ATTIVO
+          NOT_REFUNDABLE: DISATTIVATO
+          NOT_REFUNDABLE_ONLY_IBAN: METODO MANCANTE
+          NOT_REFUNDABLE_ONLY_INSTRUMENT: IBAN MANCANTE
+          UNSUBSCRIBED: DISISCRITTO
```
Comment (mastro993) [inline locales/en/index.yml:3324]:
```suggestion
          REFUNDABLE: Attiva
          NOT_REFUNDABLE: Disattivata
          NOT_REFUNDABLE_ONLY_IBAN: Metodo mancante
          NOT_REFUNDABLE_ONLY_INSTRUMENT: IBAN mancante
          UNSUBSCRIBED: Disiscritto
```

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 8]
```diff
@@ -23,133 +24,157 @@ type Props = {
...
+    <View style={styles.cardContainer} testID={"card-component"}>
+      {/* top part */}
+      <View style={styles.topCardSection}>
+        <View style={styles.bonusLogoContainer}></View>
+        <VSpacer size={8} />
+        <H1>{initiativeName}</H1>
+        <Body>{/* the ministry would go here */}</Body>
+        <VSpacer size={8} />
+        <View style={styles.bonusStatusContainer}>
+          <Badge style={styles.badge}>
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:127]: Do not use the native-base `Badge`. Instead you could easily create a badge styled with just a background color and a border radius. 
Also, the badge has differences with the original design (bigger size and bigger font)

Original design:
Screenshot 2023-03-18 alle 10 27 08

Your design:
Screenshot 2023-03-18 alle 10 27 50

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 9]
```diff
@@ -23,133 +24,157 @@ type Props = {
...
-      <View>
-        <View
-          style={styles.bottomInitiativeIcon}
-          accessible={true}
-          importantForAccessibility={"no-hide-descendants"}
-          accessibilityElementsHidden={true}
-        >
-          <Image source={bonusVacanzeWhiteLogo} style={styles.logo} />
+
+      <VSpacer size={48} />
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:148]:
```suggestion
      <VSpacer size={32} />
```

Too much spacing

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 10]
```diff
@@ -23,133 +24,157 @@ type Props = {
...
+        </View>
+        <View style={styles.alignCenter}>
+          <LabelSmall color="bluegreyDark" weight="Regular">
+            {I18n.t("idpay.initiative.details.initiativeCard.toRefund")}
+          </LabelSmall>
+          <H1 style={!isInitiativeConfigured ? styles.consumedOpacity : {}}>
+            {formatNumberAmount(toBeRepaidAmount, true)}
+          </H1>
+          <VSpacer size={4} />
+          <VSpacer size={8} />
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:173]:
```suggestion
```

To much spacing in the bottom part. This is a part of the fix that I'm suggesting, see in the above comment.

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 11]
```diff
@@ -23,133 +24,157 @@ type Props = {
...
-    position: "absolute"
-  },
-  amount: {
-    lineHeight: 32,
-    fontSize: 24,
-    color: IOColors.white,
-    ...makeFontStyleObject("Bold", undefined, "TitilliumWeb")
+    flexDirection: "row",
+    justifyContent: "space-evenly",
+    alignItems: "flex-end"
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:52]:
```suggestion
```

To much spacing in the bottom part. This is a part of the fix that I'm suggesting: removing spacers and `flex-end` automatically align items on top.

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 12]
```diff
@@ -23,133 +24,157 @@ type Props = {
...
+          <H1 style={!isInitiativeConfigured ? styles.consumedOpacity : {}}>
+            {formatNumberAmount(remainingAmount, true)}
+          </H1>
+          <VSpacer size={8} />
+          <BonusPercentageSlider
+            isGreyedOut={isInitiativeConfigured}
+            percentage={remainingBonusAmountPercentage}
+          />
+        </View>
+        <View style={styles.alignCenter}>
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:165]:
```suggestion
        </View>
        <HSpacer size={48} />
        <View style={styles.alignCenter}>
```
See comment above

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 13]
```diff
@@ -23,133 +24,157 @@ type Props = {
...
-    height: "100%",
-    position: "absolute"
-  },
-  amount: {
-    lineHeight: 32,
-    fontSize: 24,
-    color: IOColors.white,
-    ...makeFontStyleObject("Bold", undefined, "TitilliumWeb")
+    flexDirection: "row",
+    justifyContent: "space-evenly",
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:51]:
```suggestion
    justifyContent: "center",
```

This fixes the spacing in the counters part. Note that you need to manually add a spacer now, see comment below.

**Review by mastro993 — CHANGES_REQUESTED** [review]
There are lot of differences between your design and the original one. 
I made suggestions to fix the problems. Please review them and let me know.

Here is a video of the differences:

https://user-images.githubusercontent.com/6160324/226098007-7a7a4a9d-ae03-4028-864c-cbe3ca0a61a6.mov

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 14]
```diff
@@ -23,133 +24,157 @@ type Props = {
...
+      <VSpacer size={48} />
+      {/* bottom part */}
+
+      <View style={styles.bottomCardSection}>
+        <View style={styles.alignCenter}>
+          <LabelSmall color="bluegreyDark" weight="Regular">
+            {I18n.t("idpay.initiative.details.initiativeCard.availableAmount")}
+          </LabelSmall>
+          <H1 style={!isInitiativeConfigured ? styles.consumedOpacity : {}}>
+            {formatNumberAmount(remainingAmount, true)}
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:157]:
```suggestion
            {`${formatNumberAmount(remainingAmount, false)} €`}
```
Inverted currency symbol

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 15]
```diff
@@ -23,133 +24,157 @@ type Props = {
...
+            isGreyedOut={isInitiativeConfigured}
+            percentage={remainingBonusAmountPercentage}
+          />
+        </View>
+        <View style={styles.alignCenter}>
+          <LabelSmall color="bluegreyDark" weight="Regular">
+            {I18n.t("idpay.initiative.details.initiativeCard.toRefund")}
+          </LabelSmall>
+          <H1 style={!isInitiativeConfigured ? styles.consumedOpacity : {}}>
+            {formatNumberAmount(toBeRepaidAmount, true)}
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:170]:
```suggestion
            {`${formatNumberAmount(toBeRepaidAmount, false)} €`}
```
Inverted currency symbol

**Review by mastro993 — CHANGES_REQUESTED** [review]

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx (hunk 15 above)
Comment (forrest57) [reply ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx]: this and other proposed changes are adressed in https://github.com/pagopa/io-app/pull/4436/commits/dbb109705f14dcfcb7238e3c345a7878d0784bac

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 16]
```diff
@@ -23,133 +23,160 @@ type Props = {
...
+      ? (remainingAmount / amount) * 100.0
+      : 100.0;
+  return (
+    <View style={styles.cardContainer} testID={"card-component"}>
+      {/* top part */}
+      <View style={styles.topCardSection}>
+        <View style={styles.bonusLogoContainer}></View>
+        <VSpacer size={8} />
+        <H1>{initiativeName}</H1>
+        <Body>{/* the ministry would go here */}</Body>
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:127]:
```suggestion
        <LabelSmall color={"black"} weight="Regular">{/* the ministry would go here */}</LabelSmall>
```
The text in the design is smaller and lighter. `LabelSmall` with correct props should be ok.

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 17]
```diff
@@ -23,133 +23,160 @@ type Props = {
...
+        <VSpacer size={8} />
+        <View style={styles.bonusStatusContainer}>
+          <View style={styles.badge}>
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:130]: Errata corridge for my last review comment: it should be used the `IOBadge` here (also in preparation for the next DS iteration).
Please use this component instead:
```
<IOBadge
  small={true}
  text={I18n.t(
    `idpay.initiative.details.initiativeCard.statusLabels.${status}`
  )}
/>
```

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 18]
```diff
@@ -23,133 +23,160 @@ type Props = {
...
-                styles.amount
-              ]}
-            >
-              {formatNumberAmount(toBeRepaidAmount, true)}
-            </Text>
-            <H5 color="white">
-              {I18n.t("idpay.initiative.details.initiativeCard.toRefund")}
-            </H5>
-          </View>
+          <LabelSmall color="bluegreyDark">
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:138]:
```suggestion
          <LabelSmall color="bluegreyDark" fontSize="small" weight="SemiBold">
```

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 19]
```diff
@@ -23,133 +23,160 @@ type Props = {
...
+};
+const formatNumberRightSign = (amount: number) =>
+  `${formatNumberAmount(amount, false)} €`;
 const InitiativeCardComponent = (props: Props) => {
   const { initiativeName, endDate, status, amount, accrued, refunded } =
     props.initiative;
 
   const isInitiativeConfigured = status === InitiativeStatusEnum.REFUNDABLE;
   const toBeRepaidAmount = (accrued || 0) - (refunded || 0);
+  const remainingAmount = (amount || 0) - (accrued || 0);
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:113]: Screenshot 2023-03-21 alle 10 39 55

We still have problems here. From BE we have:
- `amount`: the amount still available in the initiative
- `accrued`: total amount accrued with transactions
- `refunded`: amount refunded by wire transfer

Example: initiative with 100 euros availability.

At start we have:
- `amount`: 100
- `accrued`: 0
- `refunded`: 0
- `toBeRefunded`: 0

After 30 euros of transactions (assuming 100% eligibility)
- `amount`: 70
- `accrued`: 30
- `refunded`: 0
- `toBeRefunded`: 30

After first refund:
- `amount`: 70
- `accrued`: 30
- `refunded`: 30
- `toBeRefunded`: 0

In other words, the amount to display is `amount`, but the total (the one used to set the maximum size of the progress bar) should be `amount + accrued`. 

```suggestion
  const totalAmount = (amount || 0) + (accrued || 0);
  const remainingAmount = (totalAmount || 0) - (accrued || 0);
```

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 20]
```diff
@@ -23,133 +23,160 @@ type Props = {
...
-              ]}
-            >
-              {formatNumberAmount(amount || 0, true)}
-            </Text>
-            <H5 color="white">
+  const dateString = formatDateAsLocal(endDate, true);
+  const remainingBonusAmountPercentage =
+    amount !== undefined && amount !== 0
+      ? (remainingAmount / amount) * 100.0
+      : 100.0;
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:119]:
```suggestion
  const remainingBonusAmountPercentage =
    totalAmount !== 0
      ? (remainingAmount / totalAmount) * 100.0
      : 100.0;
```

See comment

**Review by mastro993 — CHANGES_REQUESTED** [review]

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx (hunk 20 above)
Comment (forrest57) [reply ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx]: changes adressed in https://github.com/pagopa/io-app/pull/4436/commits/388f924eb3fa5ea2dc4ec0ab6ace56c49468831c

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx (hunk 19 above)
Comment (mastro993) [reply ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:108]:
```suggestion
  const remainingAmount = (amount || 0);
```
As per my previous comment, this should be the correct `remainingAmount`

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 21]
```diff
@@ -1,155 +1,181 @@
...
-                !isInitiativeConfigured ? styles.consumedOpacity : {},
-                styles.amount
-              ]}
-            >
-              {formatNumberAmount(amount || 0, true)}
-            </Text>
-            <H5 color="white">
-              {I18n.t(
-                "idpay.initiative.details.initiativeCard.availableAmount"
+            {/* no idea why, but this <View/> is required to correctly center the badge  */}
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:128]: This is not a comment a developer would like to see when reading your code 😆
Instead, try to find why this happens!

@dmnplb <IOBadge /> should use `alignSelf: "auto"` or it ignores any alignment given by the parent. Could you confirm?

**Review by mastro993 — CHANGES_REQUESTED** [review]

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 22]
```diff
@@ -1,155 +1,182 @@
...
+  const { initiativeName, endDate, status } = props.initiative;
+
+  const amount: number = props.initiative.amount || 0;
+  const accrued: number = props.initiative.accrued || 0;
+  const refunded: number = props.initiative.refunded || 0;
 
   const isInitiativeConfigured = status === InitiativeStatusEnum.REFUNDABLE;
-  const toBeRepaidAmount = (accrued || 0) - (refunded || 0);
+  const toBeRepaidAmount = accrued - refunded;
+  const remainingAmount = amount - accrued;
```
Comment (mastro993) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:111]:
```suggestion

```
See https://github.com/pagopa/io-app/pull/4436#discussion_r1143112840

File: ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx [hunk 23]
```diff
@@ -1,155 +1,182 @@
...
-                "idpay.initiative.details.initiativeCard.availableAmount"
-              )}
-            </H5>
-          </View>
+  const dateString = formatDateAsLocal(endDate, true);
+  const remainingBonusAmountPercentage =
+    totalAmount !== 0 ? (remainingAmount / totalAmount) * 100.0 : 100.0;
+  return (
+    <View style={styles.cardContainer} testID={"card-component"}>
+      {/* top part */}
```
Comment (francescopersico) [inline ts/features/idpay/initiative/details/components/InitiativeCardComponent.tsx:119]: What is this? Can we remove?
